lib/templates.php

```diff
@@ -75,19 +75,94 @@ function gutenberg_register_template_post_type() {
 	);
 
 	register_post_type( 'wp_template', $args );
+}
+add_action( 'init', 'gutenberg_register_template_post_type' );
+
+/**
+ * Registers block editor 'wp_theme' taxonomy.
+ */
+function gutenberg_register_wp_theme_taxonomy() {
+	if ( ! gutenberg_is_fse_theme() ) {
+		return;
+	}
 
-	$meta_args = array(
-		'object_subtype' => 'wp_template',
-		'type'           => 'string',
-		'description'    => 'The theme that provided the template, if any.',
-		'single'         => true,
-		'show_in_rest'   => true,
+	register_taxonomy(
+		'wp_theme',
+		array( 'wp_template', 'wp_template_part' ),
+		array(
+			'public'            => false,
+			'hierarchical'      => false,
+			'labels'            => array(
+				'name'          => __( 'Themes', 'gutenberg' ),
+				'singular_name' => __( 'Theme', 'gutenberg' ),
+			),
+			'query_var'         => false,
+			'rewrite'           => false,
+			'show_ui'           => false,
+			'_builtin'          => true,
+			'show_in_nav_menus' => false,
+			'show_in_rest'      => true,
+		)
 	);
+}
+add_action( 'init', 'gutenberg_register_wp_theme_taxonomy' );
 
-	register_post_type( 'wp_template', $args );
-	register_meta( 'post', 'theme', $meta_args );
+/**
+ * Migrates old meta based theme storage to taxonomy
+ * for wp_template and wp_template_part posts.
+ */
+function gutenberg_migrate_theme_meta_to_taxonomy() {
+	if ( ! gutenberg_is_fse_theme() || ( defined( 'REST_REQUEST' ) && REST_REQUEST ) || ( defined( 'DOING_AJAX' ) && DOING_AJAX ) ) {
+		return;
+	}
+
+	$terms = get_terms(
+		array(
+			'taxonomy'       => 'wp_theme',
+			'posts_per_page' => 1,
+			'hide_empty'     => false,
+			'fields'         => 'ids',
+		)
+	);
+	if ( $terms ) {
+		return;
+	}
+
+	$query = new WP_Query(
+		array(
+			'post_type'      => array( 'wp_template', 'wp_template_part' ),
+			'posts_per_page' => -1,
+			'post_status'    => array( 'publish', 'auto-draft', 'draft', 'trash' ),
+		)
+	);
+	if ( ! $query->have_posts() ) {
+		return;
+	}
+
+	while ( $query->have_posts() ) {
+		$query->the_post();
+		$post  = get_post();
+		$terms = array();
+
+		$theme = get_post_meta( $post->ID, 'theme', true );
+		if ( ! $theme ) {
+			$theme = wp_get_theme()->get_stylesheet();
+		}
+		$terms[] = $theme;
+
+		if ( 'auto-draft' === $post->post_status ) {
+			$terms[] = 'wp_file_based';
+		}
+
+		wp_set_post_terms( $post->ID, $terms, 'wp_theme', true );
+	}
+	wp_reset_postdata();
+
+	global $wp;
+	wp_redirect( add_query_arg( $wp->query_vars, home_url( $wp->request ) ) );
+	exit;
 }
-add_action( 'init', 'gutenberg_register_template_post_type' );
+add_action( 'init', 'gutenberg_migrate_theme_meta_to_taxonomy' );
```

Comment:
I don't think we really need a migration process here... 
FSE themes are extremely experimental and there's just 1 theme in the repo with very few installs - mostly by people working in Gutenberg.
If this migration function is removed, nothing breaks... worst case scenario there are a couple more entries in the db. Which is OK, and preferable to shipping code that will eventually have to be removed IMO

Reply:
Agree!
We are pushing so many breaking changes around Full Site Editing, that I don't really see the need of a migration for these meta.